swarm/api/http/error.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+/*
+error handling for the http package
+*/
+
+package http
+
+import (
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/swarm/api/http/tmpl"
+)
+
+var templateMap map[int]*template.Template
+
+type ErrorParams struct {
+	Msg       string
+	Code      int
+	Timestamp string
+	template  *template.Template
+	Details   template.HTML
+}
+
+func init() {
+	initErrHandling()
+}
+
+func initErrHandling() {
+  genErrPage    := tmpl.GetGenericErrorPage()
+  notFoundPage  := tmpl.GetNotFoundErrorPage()
+	tnames := map[int]string{
+		0:   genErrPage, //default
+		400: genErrPage,
+		404: notFoundPage,
+		500: genErrPage,
+	}
+	templateMap = make(map[int]*template.Template)
+	for code, tname := range tnames {
+		templateMap[code] = template.Must(template.New(fmt.Sprint("%d",code)).Parse(tname))
+	}
+}
+
+func ShowError(w http.ResponseWriter, r *http.Request, msg string, code int) {
+	params := new(ErrorParams)
+	params.Code = code
+	params.Msg = msg
+	params.Timestamp = time.Now().Format(time.RFC1123)
+	params.template = getTemplate(code)
+
+	respond(w, r, params)
+}
+
+func respond(w http.ResponseWriter, r *http.Request, params *ErrorParams) {
+	accept := r.Header.Get("Accept")
+
+	switch {
+	case accept == "text/html":
+		respondHtml(w, params)
+	case accept == "application/xhtml+xml":
+		respondHtml(w, params)
+	case accept == "application/xml":
+		respondHtml(w, params)
+	case accept == "application/json":
+		respondJson(w, params)
+	default:
+		respondHtml(w, params)
+	}
+}
+
+func respondHtml(w http.ResponseWriter, params *ErrorParams) {
+	if params.Code == http.StatusMultipleChoices {
```

Comment:
Ah - this is a remnant of an attempt at addressing https://github.com/ethereum/go-ethereum/issues/14872. If there are no strong objections, I actually would like to keep this HTTP status just for that, as to me it seems quite appropriate (I personally dislike the general custom of returning unspecified HTTP error codes like 400 or 500 for everything). This would require a separate template too.